From 161d606cd119b6c16ffef845114536fbf12ab0a0 Mon Sep 17 00:00:00 2001
From: Matthew Yu <matthewmingyu@google.com>
Date: Fri, 15 Aug 2025 16:45:10 -0700
Subject: [PATCH] Verify calling user on mms downloads

Bug: 305710989
Test: manually tested sending/downloading MMS with the primary user. I created a secondary user and also verified that sending/downloading MMS's work on the secondary user's account
Flag: EXEMPT bugfix

Change-Id: I01782ba1358b64d49ccbab0a106b60deaa9262fa
Merged-In: I01782ba1358b64d49ccbab0a106b60deaa9262fa
(cherry picked from commit 0b8c94aa0a37c83d03f6c442c8d111fe8f49bbac)
---
 src/com/android/mms/service/DownloadRequest.java |  7 ++++---
 src/com/android/mms/service/MmsRequest.java      |  3 ++-
 src/com/android/mms/service/MmsService.java      | 15 +++++++++++----
 3 files changed, 17 insertions(+), 8 deletions(-)

diff --git a/src/com/android/mms/service/DownloadRequest.java b/src/com/android/mms/service/DownloadRequest.java
index d81b384..bc538f6 100644
--- a/src/com/android/mms/service/DownloadRequest.java
+++ b/src/com/android/mms/service/DownloadRequest.java
@@ -58,16 +58,17 @@ public class DownloadRequest extends MmsRequest {
     private final String mLocationUrl;
     private final PendingIntent mDownloadedIntent;
     private final Uri mContentUri;
+    private final int mCallingUser;
 
     public DownloadRequest(RequestManager manager, int subId, String locationUrl,
-            Uri contentUri, PendingIntent downloadedIntent, String creator,
+            Uri contentUri, PendingIntent downloadedIntent, int callingUser, String creator,
             Bundle configOverrides, Context context, long messageId, MmsStats mmsStats) {
         super(manager, subId, creator, configOverrides, context, messageId, mmsStats);
         mLocationUrl = locationUrl;
         mDownloadedIntent = downloadedIntent;
         mContentUri = contentUri;
+	mCallingUser = callingUser;
     }
-
     @Override
     protected byte[] doHttp(Context context, MmsNetworkManager netMgr, ApnSettings apn)
             throws MmsHttpException {
@@ -252,7 +253,7 @@ public class DownloadRequest extends MmsRequest {
      */
     @Override
     protected boolean transferResponse(Intent fillIn, final byte[] response) {
-        return mRequestManager.writePduToContentUri(mContentUri, response);
+        return mRequestManager.writePduToContentUri(mContentUri, response, mCallingUser);
     }
 
     @Override
diff --git a/src/com/android/mms/service/MmsRequest.java b/src/com/android/mms/service/MmsRequest.java
index a27ff8a..8ebb817 100644
--- a/src/com/android/mms/service/MmsRequest.java
+++ b/src/com/android/mms/service/MmsRequest.java
@@ -87,9 +87,10 @@ public abstract class MmsRequest {
          * Write pdu to supplied content uri
          * @param contentUri content uri to which bytes should be written
          * @param pdu pdu bytes to write
+         * @param callingUser user id of the calling app
          * @return true in case of success (else false)
          */
-        public boolean writePduToContentUri(final Uri contentUri, final byte[] pdu);
+        boolean writePduToContentUri(Uri contentUri, byte[] pdu, int callingUser);
     }
 
     // The reference to the pending requests manager (i.e. the MmsService)
diff --git a/src/com/android/mms/service/MmsService.java b/src/com/android/mms/service/MmsService.java
index 092873f..7299094 100644
--- a/src/com/android/mms/service/MmsService.java
+++ b/src/com/android/mms/service/MmsService.java
@@ -355,9 +355,9 @@ public class MmsService extends Service implements MmsRequest.RequestManager {
                 return;
             }
 
-            final DownloadRequest request = new DownloadRequest(MmsService.this, subId, locationUrl,
-                    contentUri, downloadedIntent, callingPkg, mmsConfig, MmsService.this,
-                    messageId, mmsStats);
+	    final DownloadRequest request = new DownloadRequest(MmsService.this, subId, locationUrl,
+                    contentUri, downloadedIntent, callingUser, callingPkg, mmsConfig,
+                    MmsService.this, messageId, mmsStats);
 
             final String carrierMessagingServicePackage =
                     getCarrierMessagingServicePackageIfExists(subId);
@@ -1128,12 +1128,19 @@ public class MmsService extends Service implements MmsRequest.RequestManager {
      *
      * @param contentUri content provider uri to which bytes should be written
      * @param pdu        Bytes to write
+     * @param callingUser user id of the calling app
      * @return true if all bytes successfully written else false
      */
-    public boolean writePduToContentUri(final Uri contentUri, final byte[] pdu) {
+    public boolean writePduToContentUri(final Uri contentUri, final byte[] pdu, int callingUser) {
         if (contentUri == null || pdu == null) {
             return false;
         }
+        int contentUriUserID = ContentProvider.getUserIdFromUri(contentUri, UserHandle.myUserId());
+        if (callingUser != contentUriUserID) {
+            LogUtil.e("Uri belongs to a different user. contentUriUserId is: " + contentUriUserID
+                    + "and calling User ID is:" + callingUser);
+            return false;
+        }
         final Callable<Boolean> copyDownloadedPduToOutput = new Callable<Boolean>() {
             public Boolean call() {
                 ParcelFileDescriptor.AutoCloseOutputStream outStream = null;
-- 
2.34.1

